<!DOCTYPE html><html lang="ru" data-theme-enabled="1"><head><script>window.currentUser = null;</script><script>window.rateUsdToNative = 98.8745;</script><title itemprop="name">Обработка ошибок, &quot;try..catch&quot;</title><link href="https://learn.javascript.ru/pack/styles.151c2ad2b575d09af8f5.css" rel="stylesheet"><meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=yes, minimum-scale=1.0"><meta name="apple-mobile-web-app-capable" content="yes"><!-- chrome autotranslate is enabled only for "en" main version--><meta name="google" content="notranslate"><script>if (window.devicePixelRatio > 1) document.cookie = 'pixelRatio=' + window.devicePixelRatio + ';path=/;expires=Tue, 19 Jan 2038 03:14:07 GMT';</script><link href="https://fonts.googleapis.com/css?family=Open+Sans:bold,italic,bolditalic" rel="stylesheet"><link rel="apple-touch-icon-precomposed" href="https://learn.javascript.ru/img/favicon/apple-touch-icon-precomposed.png"><link rel="canonical" href="try-catch.html"><meta name="msapplication-TileColor" content="#222A2C"><meta name="msapplication-TileImage" content="/img/favicon/tileicon.png"><link rel="icon" href="https://learn.javascript.ru/img/favicon/favicon.png"><meta itemprop="image" content="https://learn.javascript.ru/img/site_preview_ru_512x512.png"><meta property="og:title" content="Обработка ошибок, &quot;try..catch&quot;"><meta property="og:image" content="https://learn.javascript.ru/img/site_preview_ru_1200x630.png"><meta property="og:image:type" content="image/png"><meta property="og:image:width" content="1200"><meta property="og:image:height" content="630"><meta property="fb:admins" content="100001562528165"><meta name="twitter:card" content="summary"><meta name="twitter:title" content="Обработка ошибок, &quot;try..catch&quot;"><meta name="twitter:site" content="@iliakan"><meta name="twitter:creator" content="@iliakan"><meta name="twitter:image" content="https://learn.javascript.ru/img/site_preview_ru_512x512.png"><link rel="prev" href="https://learn.javascript.ru/error-handling"><link rel="next" href="https://learn.javascript.ru/custom-errors"><script>window.GA_ID = "UA-2056213-16";</script><script>window.GTM_ID = 'GTM-WD2DZPG'</script><script>window.YANDEX_METRIKA_ID = 17649010;</script><script>window.GoogleAnalyticsObject="ga",window.ga=function(){window.ga.q.push(arguments)},window.ga.q=[],window.ga.l=Date.now(),ga("create",GA_ID,"auto"),window.GTM_ID&&ga("require",GTM_ID),window.currentUser?ga("set","&uid",currentUser.id):ga("set","anonymizeIp",!0),window.addEventListener("error",function(e){var r=(e.filename||e.errorUrl)+": "+(e.lineno||e.errorLine),n=e.stack||e.error&&e.error.stack;ga("send","exception",{exDescription:e.message+" "+r+" "+n,exFatal:!0})});</script><script src="https://www.google-analytics.com/analytics.js" async></script><script>ga('send', 'pageview');</script><script>window.metrika={reachGoal:function(){}},window.yandex_metrika_callbacks=[function(){try{window.metrika=new Ya.Metrika({id:YANDEX_METRIKA_ID,webvisor:!0,clickmap:!0,params:{user:window.currentUser&&window.currentUser.id}}),metrika.trackLinks({delay:150}),window.addEventListener("error",function(r){window.metrika.reachGoal("JSERROR",{src:(r.filename||r.errorUrl)+": "+(r.lineno||r.errorLine),stack:r.stack||r.error&&r.error.stack,message:r.message})})}catch(r){}}];</script><script src="https://mc.yandex.ru/metrika/watch.js" async></script><script>window.RECAPTCHA_ID = "6LfmLAEVAAAAAJMykMnf7aY8nkyTRmYi2ynx51R1";</script><script src="https://learn.javascript.ru/pack/init.f43ee4b37d6f76e4f838.js"></script><script src="https://learn.javascript.ru/pack/head.9546f64faccf13e736f6.js" defer></script><meta property="og:title" content="Обработка ошибок, &quot;try..catch&quot;"><meta property="og:type" content="article"><script src="https://learn.javascript.ru/pack/tutorial.f792acb067a354bc7904.js" defer></script><script src="https://learn.javascript.ru/pack/footer.440ca4f81c63a09affe0.js" defer></script></head><body class="no-icons"><script>window.fontTest();</script><div class="page-wrapper page-wrapper_sidebar_on"><!--[if IE]><div style="color:red;text-align:center">Извините, Internet Explorer не поддерживается, пожалуйста используйте более новый браузер.</div><![endif]--><div class="sitetoolbar sitetoolbar_tutorial"><script>window.langs = [{"code":"ar","name":"Arabic"},{"code":"az","name":"Azerbaijani"},{"code":"bg","name":"Bulgarian"},{"code":"bn","name":"Bengali"},{"code":"bs","name":"Bosnian"},{"code":"ca","name":"Catalan"},{"code":"cs","name":"Czech"},{"code":"da","name":"Danish"},{"code":"de","name":"German"},{"code":"el","name":"Greek"},{"code":"en","name":"English"},{"code":"es","name":"Spanish"},{"code":"fa","name":"Persian (Farsi)"},{"code":"fi","name":"Finnish"},{"code":"fr","name":"French"},{"code":"he","name":"Hebrew"},{"code":"hi","name":"Hindi"},{"code":"hr","name":"Croatian"},{"code":"hu","name":"Hungarian"},{"code":"hy","name":"Armenian"},{"code":"id","name":"Indonesian"},{"code":"it","name":"Italian"},{"code":"ja","name":"Japanese"},{"code":"ka","name":"Georgian"},{"code":"kk","name":"Kazakh"},{"code":"km","name":"Central Khmer"},{"code":"ko","name":"Korean"},{"code":"lt","name":"Lithuanian"},{"code":"me","name":"Montenegrin"},{"code":"ml","name":"Malayalam"},{"code":"my","name":"Burmese"},{"code":"nl","name":"Dutch"},{"code":"no","name":"Norvegian"},{"code":"pa","name":"Punjabi"},{"code":"pl","name":"Polish"},{"code":"pt","name":"Portuguese"},{"code":"ro","name":"Romanian"},{"code":"ru","name":"Russian"},{"code":"si","name":"Sinhala"},{"code":"sk","name":"Slovak"},{"code":"sl","name":"Slovenian"},{"code":"sq","name":"Albanian"},{"code":"sr","name":"Serbian"},{"code":"ta","name":"Tamil"},{"code":"te","name":"Telugu"},{"code":"test","name":"Test"},{"code":"th","name":"Thai"},{"code":"tk","name":"Turkmen"},{"code":"tr","name":"Turkish"},{"code":"uk","name":"Ukrainian"},{"code":"ur","name":"Urdu"},{"code":"uz","name":"Uzbek"},{"code":"v2","name":"v2"},{"code":"vi","name":"Vietnamese"},{"code":"zh-hant","name":"Chinese Traditional"},{"code":"zh","name":"Chinese"}];</script><script>window.lang = "ru";</script><script>{let t=navigator.languages||[];t=t.map(t=>t.toLowerCase());let o,i,n=[];for(let o of window.langs)for(let i of t)if(i===o.code||i.startsWith(o.code+"-")){n.push(o);break}if(!o&&"ru"!=lang&&"en"!=lang){n.find(t=>"en"==t.code)&&(o=`\n            According to your browser headers, you know English. Please help to <a href="https://github.com/javascript-tutorial/${lang}.javascript.info#readme">translate the tutorial</a>.\n            Thank you!\n          `,i="notify-translate-tutorial-local")}if(o){let t=`<div class="notification notification_top notification_info sitetoolbar__notification" style="display:none" id="${i}">\n          <div class="notification__content">${o}</div>\n          <button class="notification__close" title="Close"></button>\n        </div>`;document.write(t),showTopNotification()}}</script><div class="sitetoolbar__content"><div class="sitetoolbar__lang-switcher"><button class="sitetoolbar__dropdown-button" data-dropdown-toggler>RU</button><div class="sitetoolbar__dropdown-wrap"><div class="sitetoolbar__dropdown-body"><div class="sitetoolbar__lang-switcher-body"><div class="supported-langs supported-langs_toolbar"><div class="supported-langs__container"><ul class="supported-langs__list" style="height:200px"><li class="supported-langs__item"><a class="supported-langs__link" href="https://ar.javascript.info/try-catch"><span class="supported-langs__brief">AR</span><span class="supported-langs__title">عربي</span></a></li><li class="supported-langs__item"><a class="supported-langs__link" href="https://javascript.info/try-catch"><span class="supported-langs__brief">EN</span><span class="supported-langs__title">English</span></a></li><li class="supported-langs__item"><a class="supported-langs__link" href="https://es.javascript.info/try-catch"><span class="supported-langs__brief">ES</span><span class="supported-langs__title">Español</span></a></li><li class="supported-langs__item"><a class="supported-langs__link" href="https://fr.javascript.info/try-catch"><span class="supported-langs__brief">FR</span><span class="supported-langs__title">Français</span></a></li><li class="supported-langs__item"><a class="supported-langs__link" href="https://id.javascript.info/try-catch"><span class="supported-langs__brief">ID</span><span class="supported-langs__title">Indonesia</span></a></li><li class="supported-langs__item"><a class="supported-langs__link" href="https://it.javascript.info/try-catch"><span class="supported-langs__brief">IT</span><span class="supported-langs__title">Italiano</span></a></li></ul><ul class="supported-langs__list" style="height:200px"><li class="supported-langs__item"><a class="supported-langs__link" href="https://ja.javascript.info/try-catch"><span class="supported-langs__brief">JA</span><span class="supported-langs__title">日本語</span></a></li><li class="supported-langs__item"><a class="supported-langs__link" href="https://ko.javascript.info/try-catch"><span class="supported-langs__brief">KO</span><span class="supported-langs__title">한국어</span></a></li><li class="supported-langs__item supported-langs__item_current"><a class="supported-langs__link" href="try-catch.html"><span class="supported-langs__brief">RU</span><span class="supported-langs__title">Русский</span></a></li><li class="supported-langs__item"><a class="supported-langs__link" href="https://tr.javascript.info/try-catch"><span class="supported-langs__brief">TR</span><span class="supported-langs__title">Türkçe</span></a></li><li class="supported-langs__item"><a class="supported-langs__link" href="https://uk.javascript.info/try-catch"><span class="supported-langs__brief">UK</span><span class="supported-langs__title">Українська</span></a></li><li class="supported-langs__item"><a class="supported-langs__link" href="https://zh.javascript.info/try-catch"><span class="supported-langs__brief">ZH</span><span class="supported-langs__title">简体中文</span></a></li></ul></div><div class="supported-langs__text">Мы хотим сделать этот проект с открытым исходным кодом доступным для людей во всем мире. Пожалуйста, <a href="https://github.com/javascript-tutorial/translate" rel="noopener noreferrer" target="_blank">помогите нам перевести</a> это руководство на свой язык</div></div></div></div></div></div><div class="sitetoolbar__logo-wrap"><a class="sitetoolbar__link sitetoolbar__link_logo" href="https://learn.javascript.ru/"><img class="sitetoolbar__logo sitetoolbar__logo_normal" src="https://learn.javascript.ru/img/sitetoolbar__logo_ru.svg" width="171" alt="" role="presentation"/><img class="sitetoolbar__logo sitetoolbar__logo_normal sitetoolbar__logo_dark" src="https://learn.javascript.ru/img/sitetoolbar__logo_ru-white.svg" width="171" alt="" role="presentation"/><img class="sitetoolbar__logo sitetoolbar__logo_small" src="https://learn.javascript.ru/img/sitetoolbar__logo_small_ru.svg" width="80" alt="" role="presentation"/><img class="sitetoolbar__logo sitetoolbar__logo_small sitetoolbar__logo_dark" src="https://learn.javascript.ru/img/sitetoolbar__logo_small_ru-white.svg" width="80" alt="" role="presentation"/><script>Array.prototype.forEach.call(document.querySelectorAll("img.sitetoolbar__logo"),function(e){let t=document.createElement("object");t.type="image/svg+xml",t.className=e.className,t.style.cssText="left:0;top:0;position:absolute",t.onload=function(){t.onload=null,e.style.visibility="hidden"},t.data=e.src,e.parentNode.insertBefore(t,e)});</script></a></div><div class="sitetoolbar__nav-toggle-wrap"><button class="sitetoolbar__nav-toggle" type="button"></button></div><nav class="sitetoolbar__sections"><ul class="sitetoolbar__sections-list"><li class="sitetoolbar__section sitetoolbar__section_current"><a class="sitetoolbar__link" href="https://learn.javascript.ru/">Учебник</a></li><li class="sitetoolbar__section"><a class="sitetoolbar__link" href="https://learn.javascript.ru/courses">Курсы</a></li><li class="sitetoolbar__section"><a class="sitetoolbar__link" href="https://javascript.ru/forum/">Форум</a></li><li class="sitetoolbar__section"><a class="sitetoolbar__link" href="https://learn.javascript.ru/quiz">Тесты знаний</a></li><li class="sitetoolbar__section sitetoolbar__section_dropdown"><button class="sitetoolbar__dropdown-button" data-dropdown-toggler>Скринкасты</button><div class="sitetoolbar__dropdown-wrap"><div class="sitetoolbar__dropdown-body"><ul class="sitetoolbar__dropdown-items"><li class="sitetoolbar__dropdown-item"><a class="sitetoolbar__secondary-link sitetoolbar__dropdown-link" href="https://learn.javascript.ru/screencast/nodejs">Node.js</a></li><li class="sitetoolbar__dropdown-item"><a class="sitetoolbar__secondary-link sitetoolbar__dropdown-link" href="https://learn.javascript.ru/screencast/webpack">Webpack</a></li><li class="sitetoolbar__dropdown-item"><a class="sitetoolbar__secondary-link sitetoolbar__dropdown-link" href="https://learn.javascript.ru/screencast/gulp">Gulp</a></li><li class="sitetoolbar__dropdown-item"><a class="sitetoolbar__secondary-link sitetoolbar__dropdown-link" href="https://learn.javascript.ru/screencast/react">React.js</a></li><li class="sitetoolbar__dropdown-item"><a class="sitetoolbar__secondary-link sitetoolbar__dropdown-link" href="https://learn.javascript.ru/screencast/angular">Angular</a></li><li class="sitetoolbar__dropdown-item"><a class="sitetoolbar__secondary-link sitetoolbar__dropdown-link" href="https://youtu.be/W4hoc24K93E?list=PLDyvV36pndZFHXjXuwA_NywNrVQO0aQqb">Git: курс</a></li><li class="sitetoolbar__dropdown-item"><a class="sitetoolbar__secondary-link sitetoolbar__dropdown-link" href="https://youtu.be/lHacJuru1bc?list=PLDyvV36pndZEB7kWWocU4QSn-G78LoaEE">Git: разное</a></li></ul></div></div></li></ul></nav><div class="sitetoolbar__book-wrap"><a class="buy-book-button" href="https://learn.javascript.ru/ebook"><span class="buy-book-button__extra-text">Купить</span>EPUB/PDF</a></div><div class="sitetoolbar__login-wrap"><button class="sitetoolbar__login sitetoolbar__login_unready" data-action-login></button></div><div class="sitetoolbar__theme-switcher"><div class="theme-changer"><label class="theme-changer__label" for="theme-changer-input" data-tooltip="Сменить тему оформления"><input class="theme-changer__input" type="checkbox" id="theme-changer-input" data-theme-changer="data-theme-changer"/><span class="theme-changer__icon theme-changer__icon_light-theme"></span><span class="theme-changer__icon theme-changer__icon_dark-theme"></span></label></div></div><div class="sitetoolbar__search-wrap"><div class="sitetoolbar__search-content"><button class="sitetoolbar__search-toggle" type="button"></button><form class="sitetoolbar__search" method="GET" action="https://learn.javascript.ru/search"><div class="sitetoolbar__search-input"><div class="text-input"><input class="text-input__control" name="query" placeholder="Искать на Javascript.ru" required="required" type="text"/></div><button class="sitetoolbar__find" type="submit">Найти</button></div></form></div></div></div><div class="tablet-menu"><div class="tablet-menu__line"><div class="tablet-menu__content"><select class="tablet-menu__nav input-select input-select input-select_small" onchange="if(this.value) window.location.href=this.value"><option value="/" selected>Учебник</option><option value="/courses">Курсы</option><option value="https://javascript.ru/forum/">Форум</option><option value="/quiz">Тесты знаний</option></select></div></div><div class="tablet-menu__line"><div class="tablet-menu__content"><form class="tablet-menu-search" action="https://learn.javascript.ru/search/"><input class="tablet-menu-search__input" type="search" name="query" placeholder="Поиск в учебнике" required="required"/><button class="tablet-menu-search__button" type="submit" name="type" value="articles">Поиск</button></form></div></div><div class="tablet-menu__line"><div class="tablet-menu__content"><a class="map" href="https://learn.javascript.ru/tutorial/map" data-action="tutorial-map"><span class="map__text">Карта учебника</span></a></div></div><div class="tablet-menu__line"><div class="tablet-menu__content"><div class="theme-changer theme-changer_tablet-menu theme-changer_has-label"><label class="theme-changer__label" for="theme-changer-input-tablet" data-tooltip="Сменить тему оформления"><input class="theme-changer__input" type="checkbox" id="theme-changer-input-tablet" data-theme-changer="data-theme-changer"/><span class="theme-changer__icon theme-changer__icon_light-theme"></span><span class="theme-changer__icon theme-changer__icon_dark-theme"></span><span class="theme-changer__label-text theme-changer__label-text_light-theme">Светлая тема</span><span class="theme-changer__label-text theme-changer__label-text_dark-theme">Тёмная тема</span></label></div></div></div><div class="tablet-menu__line"><div class="tablet-menu__content"><div class="share-icons"><span class="share-icons__title">Поделиться</span><a class="share share_tw" href="https://twitter.com/share?url=https%3A%2F%2Flearn.javascript.ru%2Ftry-catch" rel="nofollow"></a><a class="share share_fb" href="https://www.facebook.com/sharer/sharer.php?s=100&amp;p%5Burl%5D=https%3A%2F%2Flearn.javascript.ru%2Ftry-catch" rel="nofollow"></a><a class="share share_vk" href="https://vkontakte.ru/share.php?url=https%3A%2F%2Flearn.javascript.ru%2Ftry-catch" rel="nofollow"></a></div></div></div><div class="tablet-menu__line"><div class="tablet-menu__content"><select class="tablet-menu__nav input-select input-select input-select_small" onchange="if(this.value) window.location.href=this.value"><option value="https://ar.javascript.info/try-catch">عربي</option><option value="https://javascript.info/try-catch">English</option><option value="https://es.javascript.info/try-catch">Español</option><option value="https://fr.javascript.info/try-catch">Français</option><option value="https://id.javascript.info/try-catch">Indonesia</option><option value="https://it.javascript.info/try-catch">Italiano</option><option value="https://ja.javascript.info/try-catch">日本語</option><option value="https://ko.javascript.info/try-catch">한국어</option><option value="https://learn.javascript.ru/try-catch" selected>Русский</option><option value="https://tr.javascript.info/try-catch">Türkçe</option><option value="https://uk.javascript.info/try-catch">Українська</option><option value="https://zh.javascript.info/try-catch">简体中文</option></select></div></div></div><progress class="tutorial-progress" data-sticky value="73" max="93" data-tooltip="Урок 73 из 93"></progress></div><div class="page page_sidebar_on page_inner_padding"><script>if(localStorage.noSidebar){document.querySelector(".page").classList.remove("page_sidebar_on");let e=document.querySelector(".page-wrapper");e&&e.classList.remove("page-wrapper_sidebar_on")}setTimeout(function(){document.querySelector(".page").classList.add("page_sidebar-animation-on")});</script><div class="page__inner"><main class="main main_width-limit"><header class="main__header"><div class="main__header-inner"><div class="main__header-group"><ol class="breadcrumbs"><li class="breadcrumbs__item breadcrumbs__item_home"><a class="breadcrumbs__link" href="https://learn.javascript.ru/"><span class="breadcrumbs__hidden-text">Учебник</span></a></li><li class="breadcrumbs__item" id="breadcrumb-1"><a class="breadcrumbs__link" href="https://learn.javascript.ru/js"><span>Язык JavaScript</span></a></li><li class="breadcrumbs__item" id="breadcrumb-2"><a class="breadcrumbs__link" href="https://learn.javascript.ru/error-handling"><span>Обработка ошибок</span></a></li><script type="application/ld+json">{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Учебник","item":"https://learn.javascript.ru/"},{"@type":"ListItem","position":2,"name":"Язык JavaScript","item":"https://learn.javascript.ru/js"},{"@type":"ListItem","position":3,"name":"Обработка ошибок","item":"https://learn.javascript.ru/error-handling"}]}</script></ol><div class="updated-at" data-tooltip="Последнее обновление: 14-го декабря 2021"><div class="updated-at__content">14-го декабря 2021</div></div></div><h1 class="main__header-title">Обработка ошибок, &quot;try..catch&quot;</h1></div></header><div class="content"><article class="formatted" itemscope itemtype="http://schema.org/TechArticle"><meta itemprop="name" content="Обработка ошибок, &quot;try..catch&quot;"><div itemprop="author" itemscope itemtype="http://schema.org/Person"><meta itemprop="email" content="iliakan@gmail.com"><meta itemprop="name" content="Ilya Kantor"></div><div itemprop="articleBody"><p>Неважно, насколько мы хороши в программировании, иногда наши скрипты содержат ошибки. Они могут возникать из-за наших промахов, неожиданного ввода пользователя, неправильного ответа сервера и по тысяче других причин.</p>
<p>Обычно скрипт в случае ошибки «падает» (сразу же останавливается), с выводом ошибки в консоль.</p>
<p>Но есть синтаксическая конструкция <code>try..catch</code>, которая позволяет «ловить» ошибки и вместо падения делать что-то более осмысленное.</p>
<h2><a class="main__anchor" name="sintaksis-try-catch" href="try-catch.html#sintaksis-try-catch">Синтаксис «try…catch»</a></h2><p>Конструкция <code>try..catch</code> состоит из двух основных блоков: <code>try</code>, и затем <code>catch</code>:</p>
<div id="7ttjr9cjp4" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {

  // код...

} catch (err) {

  // обработка ошибки

}</code></pre>
        </div>
      </div>
      
      </div><p>Работает она так:</p>
<ol>
<li>Сначала выполняется код внутри блока <code>try {...}</code>.</li>
<li>Если в нём нет ошибок, то блок <code>catch(err)</code> игнорируется: выполнение доходит до конца <code>try</code> и потом далее, полностью пропуская <code>catch</code>.</li>
<li>Если же в нём возникает ошибка, то выполнение <code>try</code> прерывается, и поток управления переходит в начало <code>catch(err)</code>. Переменная <code>err</code> (можно использовать любое имя) содержит объект ошибки с подробной информацией о произошедшем.</li>
</ol>
<figure><div class="image" style="width:514px">
      <div class="image__ratio" style="padding-top:79.96108949416343%"></div>
      <object type="image/svg+xml" data="https://learn.javascript.ru/article/try-catch/try-catch-flow.svg" width="514" height="411" class="image__image" data-use-theme>
        <img src="https://learn.javascript.ru/article/try-catch/try-catch-flow.svg" alt="" width="514" height="411">
      </object>
      </div></figure><p>Таким образом, при ошибке в блоке <code>try {…}</code> скрипт не «падает», и мы получаем возможность обработать ошибку внутри <code>catch</code>.</p>
<p>Давайте рассмотрим примеры.</p>
<ul>
<li>
<p>Пример без ошибок: выведет  <code>alert</code> <code>(1)</code> и <code>(2)</code>:</p>
<div id="1nkjg2ttd2" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:6,&quot;cols&quot;:[{&quot;start&quot;:33,&quot;end&quot;:40}]},{&quot;start&quot;:2,&quot;cols&quot;:[{&quot;start&quot;:33,&quot;end&quot;:40}]}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {

  alert('Начало блока try');  // (1) &lt;--

  // ...код без ошибок

  alert('Конец блока try');   // (2) &lt;--

} catch(err) {

  alert('Catch игнорируется, так как нет ошибок'); // (3)

}</code></pre>
        </div>
      </div>
      
      </div></li>
<li>
<p>Пример с ошибками: выведет <code>(1)</code> и <code>(3)</code>:</p>
<div id="r9kbz92p0w" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:10,&quot;cols&quot;:[{&quot;start&quot;:32,&quot;end&quot;:39}]},{&quot;start&quot;:4,&quot;end&quot;:4},{&quot;start&quot;:2,&quot;cols&quot;:[{&quot;start&quot;:33,&quot;end&quot;:40}]}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {

  alert('Начало блока try');  // (1) &lt;--

  lalala; // ошибка, переменная не определена!

  alert('Конец блока try (никогда не выполнится)');  // (2)

} catch(err) {

  alert(`Возникла ошибка!`); // (3) &lt;--

}</code></pre>
        </div>
      </div>
      
      </div></li>
</ul>
<div class="important important_warn">
            <div class="important__header"><span class="important__type"><code>try..catch</code> работает только для ошибок, возникающих во время выполнения кода</span></div>
            <div class="important__content"><p>Чтобы <code>try..catch</code> работал, код должен быть выполнимым. Другими словами, это должен быть корректный JavaScript-код.</p>
<p>Он не сработает, если код синтаксически неверен, например, содержит несовпадающее количество фигурных скобок:</p>
<div id="73xmzca0nt" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  {{{{{{{{{{{{
} catch(e) {
  alert(&quot;Движок не может понять этот код, он некорректен&quot;);
}</code></pre>
        </div>
      </div>
      
      </div><p>JavaScript-движок сначала читает код, а затем исполняет его. Ошибки, которые возникают во время фазы чтения, называются ошибками парсинга. Их нельзя обработать (изнутри этого кода), потому что движок не понимает код.</p>
<p>Таким образом, <code>try..catch</code> может обрабатывать только ошибки, которые возникают в корректном коде. Такие ошибки называют «ошибками во время выполнения», а иногда «исключениями».</p>
</div></div>
<div class="important important_warn">
            <div class="important__header"><span class="important__type"><code>try..catch</code> работает синхронно</span></div>
            <div class="important__content"><p>Исключение, которое произойдёт в коде, запланированном «на будущее», например в <code>setTimeout</code>, <code>try..catch</code> не поймает:</p>
<div id="y25rt3mcvc" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  setTimeout(function() {
    noSuchVariable; // скрипт упадёт тут
  }, 1000);
} catch (e) {
  alert( &quot;не сработает&quot; );
}</code></pre>
        </div>
      </div>
      
      </div><p>Это потому, что функция выполняется позже, когда движок уже покинул конструкцию <code>try..catch</code>.</p>
<p>Чтобы поймать исключение внутри запланированной функции, <code>try..catch</code> должен находиться внутри самой этой функции:</p>
<div id="inlbqljut3" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>setTimeout(function() {
  try {
    noSuchVariable; // try..catch обрабатывает ошибку!
  } catch {
    alert( &quot;ошибка поймана!&quot; );
  }
}, 1000);</code></pre>
        </div>
      </div>
      
      </div></div></div>
<h2><a class="main__anchor" name="obekt-oshibki" href="try-catch.html#obekt-oshibki">Объект ошибки</a></h2><p>Когда возникает ошибка, JavaScript генерирует объект, содержащий её детали. Затем этот объект передаётся как аргумент в блок <code>catch</code>:</p>
<div id="1me921eaul" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  // ...
} catch(err) { // &lt;-- объект ошибки, можно использовать другое название вместо err
  // ...
}</code></pre>
        </div>
      </div>
      
      </div><p>Для всех встроенных ошибок этот объект имеет два основных свойства:</p>
<dl>
<dt><code>name</code></dt>
<dd>Имя ошибки. Например, для неопределённой переменной это <code>&quot;ReferenceError&quot;</code>.</dd>
<dt><code>message</code></dt>
<dd>Текстовое сообщение о деталях ошибки.</dd>
</dl>
<p>В большинстве окружений доступны и другие, нестандартные свойства. Одно из самых широко используемых и поддерживаемых – это:</p>
<dl>
<dt><code>stack</code></dt>
<dd>Текущий стек вызова: строка, содержащая информацию о последовательности вложенных вызовов, которые привели к ошибке. Используется в целях отладки.</dd>
</dl>
<p>Например:</p>
<div id="o7y76j7ed8" data-trusted="0" class="code-example" data-highlight="[{&quot;start&quot;:1,&quot;end&quot;:1}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  lalala; // ошибка, переменная не определена!
} catch(err) {
  alert(err.name); // ReferenceError
  alert(err.message); // lalala is not defined
  alert(err.stack); // ReferenceError: lalala is not defined at (...стек вызовов)

  // Можем также просто вывести ошибку целиком
  // Ошибка приводится к строке вида &quot;name: message&quot;
  alert(err); // ReferenceError: lalala is not defined
}</code></pre>
        </div>
      </div>
      
      </div><h2><a class="main__anchor" name="blok-catch-bez-peremennoy" href="try-catch.html#blok-catch-bez-peremennoy">Блок «catch» без переменной</a></h2><div class="important important_warn">
            <div class="important__header"><span class="important__type">Новая возможность</span></div>
            <div class="important__content">
            Эта возможность была добавлена в язык недавно.
            В старых браузерах может понадобиться <a href="https://learn.javascript.ru/polyfills">полифил</a>.
            </div></div>
<p>Если нам не нужны детали ошибки, в <code>catch</code> можно её пропустить:</p>
<div id="dto1t61400" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  // ...
} catch { //  &lt;-- без (err)
  // ...
}</code></pre>
        </div>
      </div>
      
      </div><h2><a class="main__anchor" name="ispolzovanie-try-catch" href="try-catch.html#ispolzovanie-try-catch">Использование «try…catch»</a></h2><p>Давайте рассмотрим реальные случаи использования <code>try..catch</code>.</p>
<p>Как мы уже знаем, JavaScript поддерживает метод <a href="https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse">JSON.parse(str)</a> для чтения JSON.</p>
<p>Обычно он используется для декодирования данных, полученных по сети, от сервера или из другого источника.</p>
<p>Мы получаем их и вызываем <code>JSON.parse</code> вот так:</p>
<div id="yvf2ixoi5n" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:2,&quot;end&quot;:2}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>let json = '{&quot;name&quot;:&quot;John&quot;, &quot;age&quot;: 30}'; // данные с сервера

let user = JSON.parse(json); // преобразовали текстовое представление в JS-объект

// теперь user - объект со свойствами из строки
alert( user.name ); // John
alert( user.age );  // 30</code></pre>
        </div>
      </div>
      
      </div><p>Вы можете найти более детальную информацию о JSON в главе <a href="https://learn.javascript.ru/json">Формат JSON, метод toJSON</a>.</p>
<p><strong>Если <code>json</code> некорректен, <code>JSON.parse</code> генерирует ошибку, то есть скрипт «падает».</strong></p>
<p>Устроит ли нас такое поведение? Конечно нет!</p>
<p>Получается, что если вдруг что-то не так с данными, то посетитель никогда (если, конечно, не откроет консоль) об этом не узнает. А люди очень не любят, когда что-то «просто падает» без всякого сообщения об ошибке.</p>
<p>Давайте используем <code>try..catch</code> для обработки ошибки:</p>
<div id="woiy25zjar" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:8,&quot;end&quot;:11},{&quot;start&quot;:4,&quot;end&quot;:4}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>let json = &quot;{ некорректный JSON }&quot;;

try {

  let user = JSON.parse(json); // &lt;-- тут возникает ошибка...
  alert( user.name ); // не сработает

} catch (e) {
  // ...выполнение прыгает сюда
  alert( &quot;Извините, в данных ошибка, мы попробуем получить их ещё раз.&quot; );
  alert( e.name );
  alert( e.message );
}</code></pre>
        </div>
      </div>
      
      </div><p>Здесь мы используем блок <code>catch</code> только для вывода сообщения, но мы также можем сделать гораздо больше: отправить новый сетевой запрос, предложить посетителю альтернативный способ, отослать информацию об ошибке на сервер для логирования, … Всё лучше, чем просто «падение».</p>
<h2><a class="main__anchor" name="generatsiya-sobstvennyh-oshibok" href="try-catch.html#generatsiya-sobstvennyh-oshibok">Генерация собственных ошибок</a></h2><p>Что если <code>json</code> синтаксически корректен, но не содержит необходимого свойства <code>name</code>?</p>
<p>Например, так:</p>
<div id="zxr0oqa58k" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:5,&quot;end&quot;:5}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>let json = '{ &quot;age&quot;: 30 }'; // данные неполны

try {

  let user = JSON.parse(json); // &lt;-- выполнится без ошибок
  alert( user.name ); // нет свойства name!

} catch (e) {
  alert( &quot;не выполнится&quot; );
}</code></pre>
        </div>
      </div>
      
      </div><p>Здесь <code>JSON.parse</code> выполнится без ошибок, но на самом деле отсутствие свойства <code>name</code> для нас ошибка.</p>
<p>Для того, чтобы унифицировать обработку ошибок, мы воспользуемся оператором <code>throw</code>.</p>
<h3><a class="main__anchor" name="operator-throw" href="try-catch.html#operator-throw">Оператор «throw»</a></h3><p>Оператор <code>throw</code> генерирует ошибку.</p>
<p>Синтаксис:</p>
<div id="7tjvz88cpe" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>throw &lt;объект ошибки&gt;</code></pre>
        </div>
      </div>
      
      </div><p>Технически в качестве объекта ошибки можно передать что угодно. Это может быть даже примитив, число или строка, но всё же лучше, чтобы это был объект, желательно со свойствами <code>name</code> и <code>message</code> (для совместимости со встроенными ошибками).</p>
<p>В JavaScript есть множество встроенных конструкторов для стандартных ошибок: <code>Error</code>, <code>SyntaxError</code>, <code>ReferenceError</code>, <code>TypeError</code> и другие. Можно использовать и их для создания объектов ошибки.</p>
<p>Их синтаксис:</p>
<div id="tbm3fu0na8" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>let error = new Error(message);
// или
let error = new SyntaxError(message);
let error = new ReferenceError(message);
// ...</code></pre>
        </div>
      </div>
      
      </div><p>Для встроенных ошибок (не для любых объектов, только для ошибок), свойство <code>name</code> – это в точности имя конструктора. А свойство <code>message</code> берётся из аргумента.</p>
<p>Например:</p>
<div id="dyfhpx1f0d" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>let error = new Error(&quot; Ого, ошибка! o_O&quot;);

alert(error.name); // Error
alert(error.message); //  Ого, ошибка! o_O</code></pre>
        </div>
      </div>
      
      </div><p>Давайте посмотрим, какую ошибку генерирует <code>JSON.parse</code>:</p>
<div id="gvum9ru30w" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:3,&quot;end&quot;:3}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  JSON.parse(&quot;{ bad json o_O }&quot;);
} catch(e) {
  alert(e.name); // SyntaxError
  alert(e.message); // Unexpected token b in JSON at position 2
}</code></pre>
        </div>
      </div>
      
      </div><p>Как мы видим, это <code>SyntaxError</code>.</p>
<p>В нашем случае отсутствие свойства <code>name</code> – это ошибка, ведь пользователи должны иметь имена.</p>
<p>Сгенерируем её:</p>
<div id="1a8nls5q09" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:7,&quot;end&quot;:7}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>let json = '{ &quot;age&quot;: 30 }'; // данные неполны

try {

  let user = JSON.parse(json); // &lt;-- выполнится без ошибок

  if (!user.name) {
    throw new SyntaxError(&quot;Данные неполны: нет имени&quot;); // (*)
  }

  alert( user.name );

} catch(e) {
  alert( &quot;JSON Error: &quot; + e.message ); // JSON Error: Данные неполны: нет имени
}</code></pre>
        </div>
      </div>
      
      </div><p>В строке <code>(*)</code> оператор <code>throw</code> генерирует ошибку <code>SyntaxError</code> с сообщением <code>message</code>. Точно такого же вида, как генерирует сам JavaScript. Выполнение блока <code>try</code> немедленно останавливается, и поток управления прыгает в <code>catch</code>.</p>
<p>Теперь блок <code>catch</code> становится единственным местом для обработки всех ошибок: и для <code>JSON.parse</code> и для других случаев.</p>
<h2><a class="main__anchor" name="probros-isklyucheniya" href="try-catch.html#probros-isklyucheniya">Проброс исключения</a></h2><p>В примере выше мы использовали <code>try..catch</code> для обработки некорректных данных. А что, если в блоке <code>try {...}</code> возникнет <em>другая неожиданная ошибка</em>? Например, программная (неопределённая переменная) или какая-то ещё, а не ошибка, связанная с некорректными данными.</p>
<p>Пример:</p>
<div id="cb0a6qvctu" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>let json = '{ &quot;age&quot;: 30 }'; // данные неполны

try {
  user = JSON.parse(json); // &lt;-- забыл добавить &quot;let&quot; перед user

  // ...
} catch(err) {
  alert(&quot;JSON Error: &quot; + err); // JSON Error: ReferenceError: user is not defined
  // (не JSON ошибка на самом деле)
}</code></pre>
        </div>
      </div>
      
      </div><p>Конечно, возможно все! Программисты совершают ошибки. Даже в утилитах с открытым исходным кодом, используемых миллионами людей на протяжении десятилетий – вдруг может быть обнаружена ошибка, которая приводит к ужасным взломам.</p>
<p>В нашем случае <code>try..catch</code> предназначен для выявления ошибок, связанных с некорректными данными. Но по своей природе <code>catch</code> получает <em>все</em> свои ошибки из <code>try</code>. Здесь он получает неожиданную ошибку, но всё также показывает то же самое сообщение <code>&quot;JSON Error&quot;</code>. Это неправильно и затрудняет отладку кода.</p>
<p>К счастью, мы можем выяснить, какую ошибку мы получили, например, по её свойству <code>name</code>:</p>
<div id="xdp8i1f3pa" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:3,&quot;end&quot;:3}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  user = { /*...*/ };
} catch(e) {
  alert(e.name); // &quot;ReferenceError&quot; из-за неопределённой переменной
}</code></pre>
        </div>
      </div>
      
      </div><p>Есть простое правило:</p>
<p><strong>Блок <code>catch</code> должен обрабатывать только те ошибки, которые ему известны, и «пробрасывать» все остальные.</strong></p>
<p>Техника «проброс исключения» выглядит так:</p>
<ol>
<li>Блок <code>catch</code> получает все ошибки.</li>
<li>В блоке <code>catch(err) {...}</code> мы анализируем объект ошибки <code>err</code>.</li>
<li>Если мы не знаем как её обработать, тогда делаем <code>throw err</code>.</li>
</ol>
<p>В коде ниже мы используем проброс исключения, <code>catch</code> обрабатывает только <code>SyntaxError</code>:</p>
<div id="6ueyp8etqo" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:15,&quot;end&quot;:19},{&quot;start&quot;:9,&quot;end&quot;:9}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>let json = '{ &quot;age&quot;: 30 }'; // данные неполны
try {

  let user = JSON.parse(json);

  if (!user.name) {
    throw new SyntaxError(&quot;Данные неполны: нет имени&quot;);
  }

  blabla(); // неожиданная ошибка

  alert( user.name );

} catch(e) {

  if (e.name == &quot;SyntaxError&quot;) {
    alert( &quot;JSON Error: &quot; + e.message );
  } else {
    throw e; // проброс (*)
  }

}</code></pre>
        </div>
      </div>
      
      </div><p>Ошибка в строке <code>(*)</code> из блока <code>catch</code> «выпадает наружу» и может быть поймана другой внешней конструкцией <code>try..catch</code> (если есть), или «убьёт» скрипт.</p>
<p>Таким образом, блок <code>catch</code> фактически обрабатывает только те ошибки, с которыми он знает, как справляться, и пропускает остальные.</p>
<p>Пример ниже демонстрирует, как такие ошибки могут быть пойманы с помощью ещё одного уровня <code>try..catch</code>:</p>
<div id="9k0psc3lwr" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:17,&quot;end&quot;:17},{&quot;start&quot;:9,&quot;end&quot;:9},{&quot;start&quot;:5,&quot;end&quot;:5}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>function readData() {
  let json = '{ &quot;age&quot;: 30 }';

  try {
    // ...
    blabla(); // ошибка!
  } catch (e) {
    // ...
    if (e.name != 'SyntaxError') {
      throw e; // проброс исключения (не знаю как это обработать)
    }
  }
}

try {
  readData();
} catch (e) {
  alert( &quot;Внешний catch поймал: &quot; + e ); // поймал!
}</code></pre>
        </div>
      </div>
      
      </div><p>Здесь <code>readData</code> знает только, как обработать <code>SyntaxError</code>, тогда как внешний блок <code>try..catch</code> знает, как обработать всё.</p>
<h2><a class="main__anchor" name="try-catch-finally" href="try-catch.html#try-catch-finally">try…catch…finally</a></h2><p>Подождите, это ещё не всё.</p>
<p>Конструкция <code>try..catch</code> может содержать ещё одну секцию: <code>finally</code>.</p>
<p>Если секция есть, то она выполняется в любом случае:</p>
<ul>
<li>после <code>try</code>, если не было ошибок,</li>
<li>после <code>catch</code>, если ошибки были.</li>
</ul>
<p>Расширенный синтаксис выглядит следующим образом:</p>
<div id="jyi9gm4gld" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:4,&quot;cols&quot;:[{&quot;start&quot;:2,&quot;end&quot;:9}]},{&quot;start&quot;:2,&quot;cols&quot;:[{&quot;start&quot;:2,&quot;end&quot;:7}]},{&quot;start&quot;:0,&quot;cols&quot;:[{&quot;start&quot;:0,&quot;end&quot;:3}]}]">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
   ... пробуем выполнить код...
} catch(e) {
   ... обрабатываем ошибки ...
} finally {
   ... выполняем всегда ...
}</code></pre>
        </div>
      </div>
      
      </div><p>Попробуйте запустить такой код:</p>
<div id="h23z07gty6" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  alert( 'try' );
  if (confirm('Сгенерировать ошибку?')) BAD_CODE();
} catch (e) {
  alert( 'catch' );
} finally {
  alert( 'finally' );
}</code></pre>
        </div>
      </div>
      
      </div><p>У кода есть два пути выполнения:</p>
<ol>
<li>Если вы ответите на вопрос «Сгенерировать ошибку?» утвердительно, то <code>try -&gt; catch -&gt; finally</code>.</li>
<li>Если ответите отрицательно, то <code>try -&gt; finally</code>.</li>
</ol>
<p>Секцию <code>finally</code> часто используют, когда мы начали что-то делать и хотим завершить это вне зависимости от того, будет ошибка или нет.</p>
<p>Например, мы хотим измерить время, которое занимает функция чисел Фибоначчи <code>fib(n)</code>. Естественно, мы можем начать измерения до того, как функция начнёт выполняться и закончить после. Но что делать, если при вызове функции возникла ошибка? В частности, реализация <code>fib(n)</code> в коде ниже возвращает ошибку для отрицательных и для нецелых чисел.</p>
<p>Секция <code>finally</code> отлично подходит для завершения измерений несмотря ни на что.</p>
<p>Здесь <code>finally</code> гарантирует, что время будет измерено корректно в обеих ситуациях – и в случае успешного завершения <code>fib</code> и в случае ошибки:</p>
<div id="8xkzfbimap" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:17,&quot;end&quot;:19}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>let num = +prompt(&quot;Введите положительное целое число?&quot;, 35)

let diff, result;

function fib(n) {
  if (n &lt; 0 || Math.trunc(n) != n) {
    throw new Error(&quot;Должно быть целое неотрицательное число&quot;);
  }
  return n &lt;= 1 ? n : fib(n - 1) + fib(n - 2);
}

let start = Date.now();

try {
  result = fib(num);
} catch (e) {
  result = 0;
} finally {
  diff = Date.now() - start;
}

alert(result || &quot;возникла ошибка&quot;);

alert( `Выполнение заняло ${diff}ms` );</code></pre>
        </div>
      </div>
      
      </div><p>Вы можете это проверить, запустив этот код и введя <code>35</code> в <code>prompt</code> – код завершится нормально, <code>finally</code> выполнится после <code>try</code>. А затем введите <code>-1</code> – незамедлительно произойдёт ошибка, выполнение займёт <code>0ms</code>. Оба измерения выполняются корректно.</p>
<p>Другими словами, неважно как завершилась функция: через <code>return</code> или <code>throw</code>. Секция <code>finally</code> срабатывает в обоих случаях.</p>
<div class="important important_smart">
            <div class="important__header"><span class="important__type">Переменные внутри <code>try..catch..finally</code> локальны</span></div>
            <div class="important__content"><p>Обратите внимание, что переменные <code>result</code> и <code>diff</code> в коде выше объявлены <em>до</em> <code>try..catch</code>.</p>
<p>Если переменную объявить в блоке, например, в <code>try</code>, то она не будет доступна после него.</p>
</div></div>
<div class="important important_smart">
            <div class="important__header"><span class="important__type"><code>finally</code> и <code>return</code></span></div>
            <div class="important__content"><p>Блок <code>finally</code> срабатывает при <em>любом</em> выходе из <code>try..catch</code>, в том числе и <code>return</code>.</p>
<p>В примере ниже из <code>try</code> происходит <code>return</code>, но <code>finally</code> получает управление до того, как контроль возвращается во внешний код.</p>
<div id="eu6qo90u6b" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:8,&quot;end&quot;:8},{&quot;start&quot;:3,&quot;end&quot;:3}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>function func() {

  try {
    return 1;

  } catch (e) {
    /* ... */
  } finally {
    alert( 'finally' );
  }
}

alert( func() ); // сначала срабатывает alert из finally, а затем этот код</code></pre>
        </div>
      </div>
      
      </div></div></div>
<div class="important important_smart">
            <div class="important__header"><span class="important__type"><code>try..finally</code></span></div>
            <div class="important__content"><p>Конструкция <code>try..finally</code> без секции <code>catch</code> также полезна. Мы применяем её, когда не хотим здесь обрабатывать ошибки (пусть выпадут), но хотим быть уверены, что начатые процессы завершились.</p>
<div id="lsdoidtji2" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>function func() {
  // начать делать что-то, что требует завершения (например, измерения)
  try {
    // ...
  } finally {
    // завершить это, даже если все упадёт
  }
}</code></pre>
        </div>
      </div>
      
      </div><p>В приведённом выше коде ошибка всегда выпадает наружу, потому что тут нет блока <code>catch</code>. Но <code>finally</code> отрабатывает до того, как поток управления выйдет из функции.</p>
</div></div>
<h2><a class="main__anchor" name="globalnyy-catch" href="try-catch.html#globalnyy-catch">Глобальный catch</a></h2><div class="important important_warn">
            <div class="important__header"><span class="important__type">Зависит от окружения</span></div>
            <div class="important__content"><p>Информация из данной секции не является частью языка JavaScript.</p>
</div></div>
<p>Давайте представим, что произошла фатальная ошибка (программная или что-то ещё ужасное) снаружи <code>try..catch</code>, и скрипт упал.</p>
<p>Существует ли способ отреагировать на такие ситуации? Мы можем захотеть залогировать ошибку, показать что-то пользователю (обычно они не видят сообщение об ошибке) и т.д.</p>
<p>Такого способа нет в спецификации, но обычно окружения предоставляют его, потому что это весьма полезно. Например, в Node.js для этого есть <a href="https://nodejs.org/api/process.html#process_event_uncaughtexception"><code>process.on(&quot;uncaughtException&quot;)</code></a>. А в браузере мы можем присвоить функцию специальному свойству <a href="https://developer.mozilla.org/ru/docs/Web/API/GlobalEventHandlers/onerror">window.onerror</a>, которая будет вызвана в случае необработанной ошибки.</p>
<p>Синтаксис:</p>
<div id="uda1wlyrrg" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>window.onerror = function(message, url, line, col, error) {
  // ...
};</code></pre>
        </div>
      </div>
      
      </div><dl>
<dt><code>message</code></dt>
<dd>Сообщение об ошибке.</dd>
<dt><code>url</code></dt>
<dd>URL скрипта, в котором произошла ошибка.</dd>
<dt><code>line</code>, <code>col</code></dt>
<dd>Номера строки и столбца, в которых произошла ошибка.</dd>
<dt><code>error</code></dt>
<dd>Объект ошибки.</dd>
</dl>
<p>Пример:</p>
<div id="v4a424o0ke" data-trusted="0" class="code-example" data-demo-height="1" data-refresh="1" data-highlight="[{&quot;start&quot;:1,&quot;end&quot;:3}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="показать" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-markup"><code>&lt;script&gt;
  window.onerror = function(message, url, line, col, error) {
    alert(`${message}\n В ${line}:${col} на ${url}`);
  };

  function readData() {
    badFunc(); // Ой, что-то пошло не так!
  }

  readData();
&lt;/script&gt;</code></pre>
        </div>
      </div>
      
      </div><p>Роль глобального обработчика <code>window.onerror</code> обычно заключается не в восстановлении выполнения скрипта – это скорее всего невозможно в случае программной ошибки, а в отправке сообщения об ошибке разработчикам.</p>
<p>Существуют также веб-сервисы, которые предоставляют логирование ошибок для таких случаев, такие как <a href="https://errorception.com">https://errorception.com</a> или <a href="http://www.muscula.com">http://www.muscula.com</a>.</p>
<p>Они работают так:</p>
<ol>
<li>Мы регистрируемся в сервисе и получаем небольшой JS-скрипт (или URL скрипта) от них для вставки на страницы.</li>
<li>Этот JS-скрипт ставит свою функцию <code>window.onerror</code>.</li>
<li>Когда возникает ошибка, она выполняется и отправляет сетевой запрос с информацией о ней в сервис.</li>
<li>Мы можем войти в веб-интерфейс сервиса и увидеть ошибки.</li>
</ol>
<h2><a class="main__anchor" name="itogo" href="try-catch.html#itogo">Итого</a></h2><p>Конструкция <code>try..catch</code> позволяет обрабатывать ошибки во время исполнения кода. Она позволяет запустить код и перехватить ошибки, которые могут в нём возникнуть.</p>
<p>Синтаксис:</p>
<div id="0a6gaas7ba" data-trusted="1" class="code-example">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  // исполняем код
} catch(err) {
  // если случилась ошибка, прыгаем сюда
  // err - это объект ошибки
} finally {
  // выполняется всегда после try/catch
}</code></pre>
        </div>
      </div>
      
      </div><p>Секций <code>catch</code> или <code>finally</code> может не быть, то есть более короткие конструкции <code>try..catch</code> и <code>try..finally</code> также корректны.</p>
<p>Объекты ошибок содержат следующие свойства:</p>
<ul>
<li><code>message</code> – понятное человеку сообщение.</li>
<li><code>name</code> – строка с именем ошибки (имя конструктора ошибки).</li>
<li><code>stack</code> (нестандартное, но хорошо поддерживается) – стек на момент ошибки.</li>
</ul>
<p>Если объект ошибки не нужен, мы можем пропустить его, используя <code>catch {</code> вместо <code>catch(err) {</code>.</p>
<p>Мы можем также генерировать собственные ошибки, используя оператор <code>throw</code>. Аргументом <code>throw</code> может быть что угодно, но обычно это объект ошибки, наследуемый от встроенного класса <code>Error</code>. Подробнее о расширении ошибок см. в следующей главе.</p>
<p><em>Проброс исключения</em> – это очень важный приём обработки ошибок: блок <code>catch</code> обычно ожидает и знает, как обработать определённый тип ошибок, поэтому он должен пробрасывать дальше ошибки, о которых он не знает.</p>
<p>Даже если у нас нет <code>try..catch</code>, большинство сред позволяют настроить «глобальный» обработчик ошибок, чтобы ловить ошибки, которые «выпадают наружу». В браузере это <code>window.onerror</code>.</p>
</div></article><div class="tasks formatted"><h2 class="tasks__title" id="tasks"><a class="tasks__title-anchor main__anchor main__anchor main__anchor_noicon" href="try-catch.html#tasks">Задачи</a></h2><div class="task tasks__task"><div class="task__header"><div class="task__title-wrap"><h3 class="task__title"><a class="main__anchor" href="try-catch.html#finally-ili-prosto-kod" name="finally-ili-prosto-kod">Finally или просто код?</a></h3><a class="task__open-link" href="https://learn.javascript.ru/task/finally-or-code-after" target="_blank"></a></div><div class="task__header-note"><span class="task__importance" title="Насколько эта задача важна для освоения материала, от 1 до 5">важность: 5</span></div><div class="task__content"><div class="task__formatted"><p>Сравните два фрагмента кода.</p>
<ol>
<li>
<p>Первый использует <code>finally</code> для выполнения кода после <code>try..catch</code>:</p>
<div id="zade626wmd" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:6,&quot;end&quot;:6}]">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  начать работу
  работать
} catch (e) {
  обработать ошибку
} finally {
  очистить рабочее пространство
}</code></pre>
        </div>
      </div>
      
      </div></li>
<li>
<p>Второй фрагмент просто ставит очистку после <code>try..catch</code>:</p>
<div id="xpf7b8dgik" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:7,&quot;end&quot;:7}]">
      <div class="codebox code-example__codebox">
        
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>try {
  начать работу
  работать
} catch (e) {
  обработать ошибку
}

очистить рабочее пространство</code></pre>
        </div>
      </div>
      
      </div></li>
</ol>
<p>Нам определённо нужна очистка после работы, неважно возникли ошибки или нет.</p>
<p>Есть ли здесь преимущество в использовании <code>finally</code> или оба фрагмента кода одинаковы? Если такое преимущество есть, то дайте пример, когда оно проявляется.</p>
</div><button class="task__solution" type="button">решение</button><div class="task__answer"><div class="task__answer-content"><div class="formatted"><p>Разница в поведении станет очевидной, если рассмотреть код внутри функции.</p>
<p>Поведение будет различным, если управление каким-то образом выпрыгнет из <code>try..catch</code>.</p>
<p>Например, когда есть <code>return</code> внутри <code>try..catch</code>. Секция <code>finally</code> работает в любом случае при <em>любом</em> выходе из <code>try..catch</code>, даже через <code>return</code>: сразу после того как <code>try..catch</code> выполнится, но до того, как вызывающий код получит контроль.</p>
<div id="dc7csceaiy" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:3,&quot;end&quot;:3}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>function f() {
  try {
    alert('начало');
    return &quot;result&quot;;
  } catch (e) {
    /// ...
  } finally {
    alert('очистка!');
  }
}

f(); // очистка!</code></pre>
        </div>
      </div>
      
      </div><p>…Или когда есть <code>throw</code>, например, как здесь:</p>
<div id="6txuzzv4tb" data-trusted="1" class="code-example" data-highlight="[{&quot;start&quot;:7,&quot;end&quot;:7}]">
      <div class="codebox code-example__codebox">
        
        <div class="toolbar codebox__toolbar">
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="выполнить" data-action="run" class="toolbar__button toolbar__button_run"></a>
          </div>
          <div class="toolbar__tool">
            <a href="try-catch.html#" title="открыть в песочнице" target="_blank" data-action="edit" class="toolbar__button toolbar__button_edit"></a>
          </div>
        </div>
        <div class="codebox__code" data-code="1">
          <pre class="line-numbers language-javascript"><code>function f() {
  try {
    alert('начало');
    throw new Error(&quot;ошибка&quot;);
  } catch (e) {
    // ...
    if(&quot;не могу обработать ошибку&quot;) {
      throw e;
    }

  } finally {
    alert('очистка!')
  }
}

f(); // очистка!</code></pre>
        </div>
      </div>
      
      </div><p>Именно <code>finally</code> гарантирует очистку. Если мы просто поместим код в конце <code>f</code>, то он не выполнится в описанных ситуациях.</p>
</div></div><button class="close-button task__answer-close" type="button" title="закрыть"></button></div></div></div></div></div></div><div class="page__nav-wrap"><a class="page__nav page__nav_prev" href="https://learn.javascript.ru/error-handling" data-tooltip="Обработка ошибок"><span class="page__nav-text"><span class="page__nav-text-shortcut"></span></span><span class="page__nav-text-alternate">Предыдущий урок</span></a><a class="page__nav page__nav_next" href="https://learn.javascript.ru/custom-errors" data-tooltip="Пользовательские ошибки, расширение Error"><span class="page__nav-text"><span class="page__nav-text-shortcut"></span></span><span class="page__nav-text-alternate">Следующий урок</span></a></div><div class="article-tablet-foot tablet-only"><div class="article-tablet-foot__layout"><div class="share-icons"><span class="share-icons__title">Поделиться</span><a class="share share_tw" href="https://twitter.com/share?url=https%3A%2F%2Flearn.javascript.ru%2Ftry-catch" rel="nofollow"></a><a class="share share_fb" href="https://www.facebook.com/sharer/sharer.php?s=100&amp;p%5Burl%5D=https%3A%2F%2Flearn.javascript.ru%2Ftry-catch" rel="nofollow"></a><a class="share share_vk" href="https://vkontakte.ru/share.php?url=https%3A%2F%2Flearn.javascript.ru%2Ftry-catch" rel="nofollow"></a></div><div class="article-tablet-foot__map"><a class="map" href="https://learn.javascript.ru/tutorial/map" data-action="tutorial-map"><span class="map__text">Карта учебника</span></a></div></div></div><div class="banner-bottom"><div class="banner-bottom__text">Проводим <a href="https://learn.javascript.ru/courses">курсы по JavaScript и фреймворкам</a>.</div><button class="banner-bottom__close" data-banner-close="Courses" data-banner-close-message="Эта информация больше не будет выводиться." title="не показывать"></button></div><script>!!1&&"hideBannerCourses"in localStorage||(document.querySelector(".banner-bottom").style.display="flex");</script><div class="comments formatted" id="comments"><div class="comments__header"><h2 class="comments__header-title"><a href="try-catch.html#comments" name="comments">Комментарии</a></h2><div class="comments__read-before"><span class="comments__read-before-link">перед тем как писать…</span><div class="comments__read-before-popup"><div class="comments__read-before-popup-i"><ul><li>Если вам кажется, что в статье что-то не так - вместо комментария напишите <a href="https://github.com/javascript-tutorial/ru.javascript.info/issues/new">на GitHub</a>.</li><li>Для одной строки кода используйте тег <code>&lt;code&gt;</code>, для нескольких строк кода&nbsp;&mdash; тег <code>&lt;pre&gt;</code>, если больше 10 строк&nbsp;&mdash; ссылку на песочницу (<a href='https://plnkr.co/edit/?p=preview'>plnkr</a>, <a href='http://jsbin.com'>JSBin</a>, <a href='http://codepen.io'>codepen</a>…)</li><li>Если что-то непонятно в статье&nbsp;&mdash; пишите, что именно и с какого места.</li></ul></div></div></div></div><div id="disqus_thread"></div><script>var disqus_config = function() { if (!this.page) this.page = {}; Object.assign(this.page, {"url":"https:\/\/learn.javascript.ru\/try-catch","identifier":"\/try-catch"}); };</script><script>var disqus_shortname = "learnjavascriptru";</script><script>var disqus_enabled = true;</script></div></main></div><div class="sidebar page__sidebar sidebar sidebar_sticky-footer"><button class="sidebar__toggle" data-sidebar-toggle></button><a class="map" href="https://learn.javascript.ru/tutorial/map" data-action="tutorial-map" data-tooltip="Карта учебника"></a><div class="sidebar__inner"><div class="sidebar__content"><div class="sidebar__section"><h4 class="sidebar__section-title">Раздел</h4><nav class="sidebar__navigation"><ul class="sidebar__navigation-links"><li class="sidebar__navigation-link"><a class="sidebar__link" href="https://learn.javascript.ru/error-handling">Обработка ошибок</a></li></ul></nav></div><div class="sidebar__section"><h4 class="sidebar__section-title">Навигация по уроку</h4><nav class="sidebar__navigation"><ul class="sidebar__navigation-links"><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#sintaksis-try-catch">Синтаксис «try…catch»</a></li><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#obekt-oshibki">Объект ошибки</a></li><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#blok-catch-bez-peremennoy">Блок «catch» без переменной</a></li><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#ispolzovanie-try-catch">Использование «try…catch»</a></li><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#generatsiya-sobstvennyh-oshibok">Генерация собственных ошибок</a></li><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#probros-isklyucheniya">Проброс исключения</a></li><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#try-catch-finally">try…catch…finally</a></li><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#globalnyy-catch">Глобальный catch</a></li><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#itogo">Итого</a></li></ul></nav></div><div class="sidebar__section"><nav class="sidebar__navigation"><ul class="sidebar__navigation-links"><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#tasks">Задачи (1)</a></li><li class="sidebar__navigation-link"><a class="sidebar__link" href="try-catch.html#comments">Комментарии</a></li></ul></nav></div><div class="sidebar__section"><div class="sidebar__section-title">Поделиться</div><a class="share share_tw sidebar__share" href="https://twitter.com/share?url=https%3A%2F%2Flearn.javascript.ru%2Ftry-catch" rel="nofollow"></a><a class="share share_fb sidebar__share" href="https://www.facebook.com/sharer/sharer.php?s=100&amp;p[url]=https%3A%2F%2Flearn.javascript.ru%2Ftry-catch" rel="nofollow"></a></div><div class="sidebar__section"><a class="sidebar__link" href="https://github.com/javascript-tutorial/ru.javascript.info/blob/master/1-js/10-error-handling/1-try-catch" rel="nofollow">Редактировать на GitHub</a></div></div></div></div></div></div><div class="page-footer"><ul class="page-footer__list"><li class="page-footer__item page-footer__item_copy">©&nbsp;2007—2022&nbsp; Илья Кантор</li><li class="page-footer__item page-footer__item_about"><a class="page-footer__link" href="https://learn.javascript.ru/about">о проекте</a></li><li class="page-footer__item page-footer__item_contact"><a class="page-footer__link" href="https://learn.javascript.ru/about#contact-us">связаться с нами</a></li><li class="page-footer__item page-footer__item_terms"><a class="page-footer__link" href="https://learn.javascript.ru/terms">пользовательское соглашение</a></li><li class="page-footer__item page-footer__item_privacy"><a class="page-footer__link" href="https://learn.javascript.ru/privacy">политика конфиденциальности</a></li><li class="page-footer__item page-footer__item_slack"><a class="page-footer__slack" href="http://slack.javascript.ru">slack-чат</a></li></ul></div></body></html>